src/Altinn.Apps/AppTemplates/AspNet/Altinn.App.PlatformServices/Implementation/ProcessChangeHandler.cs

```diff
@@ -0,0 +1,426 @@
+                    CurrentTask = new ProcessElementInfo
+                    {
+                        Flow = e.ProcessInfo?.CurrentTask.Flow,
+                        AltinnTaskType = e.ProcessInfo?.CurrentTask.AltinnTaskType,
+                        ElementId = e.ProcessInfo?.CurrentTask.ElementId,
+                        Name = e.ProcessInfo?.CurrentTask.Name,
+                        Started = e.ProcessInfo?.CurrentTask.Started,
+                        Ended = e.ProcessInfo?.CurrentTask.Ended,
+                        Validated = new ValidationStatus
+                        {
+                            CanCompleteTask = e.ProcessInfo?.CurrentTask?.Validated?.CanCompleteTask ?? false,
+                            Timestamp = e.ProcessInfo?.CurrentTask?.Validated?.Timestamp
+                        }
+                    },
+
+                    StartEvent = e.ProcessInfo?.StartEvent
+                },
+                User = new PlatformUser
+                {
+                    AuthenticationLevel = e.User.AuthenticationLevel,
+                    EndUserSystemId = e.User.EndUserSystemId,
+                    OrgId = e.User.OrgId,
+                    UserId = e.User.UserId
+                }
+            };
+        }
+
+        /// <summary>
+        /// Moves instance's process to nextElement id. Returns the instance together with process events.
+        /// </summary>
+        public ProcessStateChange ProcessNext(Instance instance, string nextElementId, ClaimsPrincipal userContext)
+        {
+            if (instance.Process != null)
+            {
+                try
+                {
+                    ProcessStateChange result = new ProcessStateChange
+                    {
+                        OldProcessState = new ProcessState()
+                        {
+                            Started = instance.Process.Started,
+                            CurrentTask = instance.Process.CurrentTask,
+                            StartEvent = instance.Process.StartEvent
+                        }
+                    };
+
+                    result.Events = MoveProcessToNext(instance, nextElementId, userContext);
+                    result.NewProcessState = instance.Process;
+                    return result;
+                }
+                catch
+                {
+                    _logger.LogError($"Unable to get next for {instance.Id}");
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Assumes that nextElementId is a valid task/state
+        /// </summary>
+        private List<InstanceEvent> MoveProcessToNext(
+            Instance instance,
+            string nextElementId,
+            ClaimsPrincipal user)
+        {
+            List<InstanceEvent> events = new List<InstanceEvent>();
+
+            ProcessState previousState = JsonConvert.DeserializeObject<ProcessState>(JsonConvert.SerializeObject(instance.Process));
```

Comment:
This statement is using NewtonSoft serialization and deserialization to produce a clone of the original Process. We should replace it with a method that copies properties.

Reply:
fixed